src/crystal/syntax_highlighter.cr

```diff
@@ -0,0 +1,220 @@
+    when :COMMENT
+      render :COMMENT, token.value.to_s
+    when :NUMBER
+      render :NUMBER, token.raw
+    when :CHAR
+      render :CHAR, token.raw
+    when :SYMBOL
+      render :SYMBOL, token.raw
+    when :DELIMITER_START
+      render :STRING, token.raw
+    when :CONST, :"::"
+      render :CONST, token.to_s
+    when :IDENT
+      if last_is_def
+        render :IDENT, token.to_s
+      else
+        case token.value
+        when :def, :if, :else, :elsif, :end,
+             :class, :module, :include, :extend,
+             :while, :until, :do, :yield, :return, :unless, :next, :break, :begin,
+             :lib, :fun, :type, :struct, :union, :enum, :macro, :out, :require,
+             :case, :when, :select, :then, :of, :abstract, :rescue, :ensure, :is_a?,
+             :alias, :pointerof, :sizeof, :instance_sizeof, :offsetof, :as, :as?, :typeof, :for, :in,
+             :with, :super, :private, :asm, :nil?, :protected, :uninitialized, "new",
+             :annotation, :verbatim
+          render :KEYWORD, token.to_s
+        when :true, :false, :nil
+          render :PRIMITIVE_LITERAL, token.to_s
+        when :self
+          render :SELF, token.to_s
+        else
+          render :UNKNOWN, token.to_s
+        end
+      end
+    when :+, :-, :*, :&+, :&-, :&*, :/, ://,
+         :"=", :==, :<, :<=, :>, :>=, :!, :!=, :=~, :!~,
+         :&, :|, :^, :~, :**, :>>, :<<, :%,
+         :[], :[]?, :[]=, :<=>, :===
+      render :OPERATOR, token.to_s
+    when :UNDERSCORE
+      render :UNDERSCORE, "_"
+    else
+      render :UNKNOWN, token.to_s
+    end
+  end
+
+  private def highlight_delimiter_state(lexer, token, *, heredoc = false)
+    render_delimiter do
+      render :DELIMITER_START, token.raw unless heredoc
+      while true
+        token = lexer.next_string_token(token.delimiter_state)
+        case token.type
+        when :DELIMITER_END
+          render :DELIMITER_END, token.raw
+          break
+        when :INTERPOLATION_START
+          render_interpolation do
+            highlight_normal_state lexer, break_on_rcurly: true
+          end
+        else
+          render :DELIMITED_TOKEN, token.raw
+        end
+      end
+    end
+  end
+
+  private def highlight_string_array(lexer, token)
+    render_string_array do
+      render :STRING_ARRAY_START, token.raw
+      while true
+        consume_space_or_newline(lexer)
+        token = lexer.next_string_array_token
+        case token.type
+        when :STRING
+          render :STRING_ARRAY_TOKEN, token.raw
+        when :STRING_ARRAY_END
+          render :STRING_ARRAY_END, token.raw
+          break
+        when :EOF
+          if token.delimiter_state.kind == :string_array
+            raise "Unterminated string array literal"
+          else # == :symbol_array
+            raise "Unterminated symbol array literal"
+          end
+        else
+          raise "Bug: shouldn't happen"
+        end
+      end
+    end
+  end
+end
```

Comment:
It seems that the formatter now has a bug: it eats curly braces at the end of lines

![image](https://user-images.githubusercontent.com/371383/144478392-b52be5ca-4c53-4bf2-b7cd-e0fb54b48e7c.png)

<details>

```diff
--- docsbefore/Array.html	2021-12-02 18:42:57.652031756 +0100
+++ docsafter/Array.html	2021-12-02 19:00:02.976900673 +0100
@@ -3752 +3752 @@
-<pre><code class="language-crystal">set <span class="o">=</span> <span class="t">Set</span>{<span class="n">1</span>, <span class="n">2</span>, <span class="n">3</span>} <span class="c"># =&gt; Set{1, 2, 3}</span>
+<pre><code class="language-crystal">set <span class="o">=</span> <span class="t">Set</span>{<span class="n">1</span>, <span class="n">2</span>, <span class="n">3</span> <span class="c"># =&gt; Set{1, 2, 3}</span>
@@ -5835 +5835 @@
-<pre><code class="language-crystal"><span class="t">Array</span>.<span class="k">new</span>(<span class="n">3</span>) { <span class="o">|</span>i<span class="o">|</span> (i <span class="o">+</span> <span class="n">1</span>) <span class="o">**</span> <span class="n">2</span> } <span class="c"># =&gt; [1, 4, 9]</span>
+<pre><code class="language-crystal"><span class="t">Array</span>.<span class="k">new</span>(<span class="n">3</span>) { <span class="o">|</span>i<span class="o">|</span> (i <span class="o">+</span> <span class="n">1</span>) <span class="o">**</span> <span class="n">2</span>  <span class="c"># =&gt; [1, 4, 9]</span>
@@ -5837 +5837 @@
-ary <span class="o">=</span> <span class="t">Array</span>.<span class="k">new</span>(<span class="n">3</span>) { [<span class="n">1</span>] }
+ary <span class="o">=</span> <span class="t">Array</span>.<span class="k">new</span>(<span class="n">3</span>) { [<span class="n">1</span>] 
@@ -6851 +6851 @@
-a.fill(<span class="n">2</span>, <span class="n">2</span>) { <span class="o">|</span>i<span class="o">|</span> i <span class="o">*</span> i } <span class="c"># =&gt; [1, 2, 4, 9, 5, 6]</span></code></pre>
+a.fill(<span class="n">2</span>, <span class="n">2</span>) { <span class="o">|</span>i<span class="o">|</span> i <span class="o">*</span> i  <span class="c"># =&gt; [1, 2, 4, 9, 5, 6]</span></code></pre>
@@ -6877 +6877 @@
-a.fill(<span class="n">2</span>) { <span class="o">|</span>i<span class="o">|</span> i <span class="o">*</span> i } <span class="c"># =&gt; [1, 2, 4, 9]</span></code></pre>
+a.fill(<span class="n">2</span>) { <span class="o">|</span>i<span class="o">|</span> i <span class="o">*</span> i  <span class="c"># =&gt; [1, 2, 4, 9]</span></code></pre>
@@ -6901 +6901 @@
-a.fill(<span class="n">2</span>..<span class="n">3</span>) { <span class="o">|</span>i<span class="o">|</span> i <span class="o">*</span> i } <span class="c"># =&gt; [1, 2, 4, 9, 5, 6]</span></code></pre>
+a.fill(<span class="n">2</span>..<span class="n">3</span>) { <span class="o">|</span>i<span class="o">|</span> i <span class="o">*</span> i  <span class="c"># =&gt; [1, 2, 4, 9, 5, 6]</span></code></pre>
@@ -7204 +7204 @@
-results <span class="o">=</span> gems.map_with_index { <span class="o">|</span>gem, i<span class="o">|</span> <span class="s">&quot;</span><span class="i">#{</span>i<span class="i">}</span><span class="s">: </span><span class="i">#{</span>gem<span class="i">}</span><span class="s">&quot;</span> }
+results <span class="o">=</span> gems.map_with_index { <span class="o">|</span>gem, i<span class="o">|</span> <span class="s">&quot;</span><span class="i">#{</span>i<span class="i">}</span><span class="s">: </span><span class="i">#{</span>gem<span class="i">}</span><span class="s">&quot;</span> 
@@ -7288,2 +7288,2 @@
-a.pop { <span class="s">&quot;Testing&quot;</span> } <span class="c"># =&gt; 1</span>
-a.pop { <span class="s">&quot;Testing&quot;</span> } <span class="c"># =&gt; &quot;Testing&quot;</span></code></pre>
+a.pop { <span class="s">&quot;Testing&quot;</span>  <span class="c"># =&gt; 1</span>
+a.pop { <span class="s">&quot;Testing&quot;</span>  <span class="c"># =&gt; &quot;Testing&quot;</span></code></pre>
@@ -7455 +7455 @@
-ary.reject! { <span class="o">|</span>x<span class="o">|</span> x <span class="o">&gt;</span> <span class="n">3</span> }
+ary.reject! { <span class="o">|</span>x<span class="o">|</span> x <span class="o">&gt;</span> <span class="n">3</span> 
@@ -7636 +7636 @@
-ary.select! { <span class="o">|</span>x<span class="o">|</span> x <span class="o">&gt;</span> <span class="n">3</span> }
+ary.select! { <span class="o">|</span>x<span class="o">|</span> x <span class="o">&gt;</span> <span class="n">3</span> 
@@ -7747 +7747 @@
-a.shift { <span class="s">&quot;empty!&quot;</span> } <span class="c"># =&gt; &quot;a&quot;</span>
+a.shift { <span class="s">&quot;empty!&quot;</span>  <span class="c"># =&gt; &quot;a&quot;</span>
@@ -7749 +7749 @@
-a.shift { <span class="s">&quot;empty!&quot;</span> } <span class="c"># =&gt; &quot;empty!&quot;</span>
+a.shift { <span class="s">&quot;empty!&quot;</span>  <span class="c"># =&gt; &quot;empty!&quot;</span>
@@ -7900 +7900 @@
-b <span class="o">=</span> a.sort { <span class="o">|</span>a, b<span class="o">|</span> b <span class="o">&lt;=&gt;</span> a }
+b <span class="o">=</span> a.sort { <span class="o">|</span>a, b<span class="o">|</span> b <span class="o">&lt;=&gt;</span> a 
@@ -7962 +7962 @@
-a.sort! { <span class="o">|</span>a, b<span class="o">|</span> b <span class="o">&lt;=&gt;</span> a }
+a.sort! { <span class="o">|</span>a, b<span class="o">|</span> b <span class="o">&lt;=&gt;</span> a 
@@ -7992 +7992 @@
-b <span class="o">=</span> a.sort_by { <span class="o">|</span>word<span class="o">|</span> word.size }
+b <span class="o">=</span> a.sort_by { <span class="o">|</span>word<span class="o">|</span> word.size 
@@ -8023 +8023 @@
-a.sort_by! { <span class="o">|</span>word<span class="o">|</span> word.size }
+a.sort_by! { <span class="o">|</span>word<span class="o">|</span> word.size 
@@ -8056 +8056 @@
-<pre><code class="language-crystal">{<span class="n">1</span>, <span class="n">2</span>, <span class="n">3</span>}.to_a <span class="c"># =&gt; [1, 2, 3]</span></code></pre>
+<pre><code class="language-crystal">{<span class="n">1</span>, <span class="n">2</span>, <span class="n">3</span>.to_a <span class="c"># =&gt; [1, 2, 3]</span></code></pre>
@@ -8269,2 +8269,2 @@
-<pre><code class="language-crystal">a <span class="o">=</span> [{<span class="s">&quot;student&quot;</span>, <span class="s">&quot;sam&quot;</span>}, {<span class="s">&quot;student&quot;</span>, <span class="s">&quot;george&quot;</span>}, {<span class="s">&quot;teacher&quot;</span>, <span class="s">&quot;matz&quot;</span>}]
-a.uniq { <span class="o">|</span>s<span class="o">|</span> s[<span class="n">0</span>] } <span class="c"># =&gt; [{&quot;student&quot;, &quot;sam&quot;}, {&quot;teacher&quot;, &quot;matz&quot;}]</span>
+<pre><code class="language-crystal">a <span class="o">=</span> [{<span class="s">&quot;student&quot;</span>, <span class="s">&quot;sam&quot;</span>, {<span class="s">&quot;student&quot;</span>, <span class="s">&quot;george&quot;</span>, {<span class="s">&quot;teacher&quot;</span>, <span class="s">&quot;matz&quot;</span>]
+a.uniq { <span class="o">|</span>s<span class="o">|</span> s[<span class="n">0</span>]  <span class="c"># =&gt; [{&quot;student&quot;, &quot;sam&quot;}, {&quot;teacher&quot;, &quot;matz&quot;}]</span>
@@ -8317,2 +8317,2 @@
-<pre><code class="language-crystal">a <span class="o">=</span> [{<span class="s">&quot;student&quot;</span>, <span class="s">&quot;sam&quot;</span>}, {<span class="s">&quot;student&quot;</span>, <span class="s">&quot;george&quot;</span>}, {<span class="s">&quot;teacher&quot;</span>, <span class="s">&quot;matz&quot;</span>}]
-a.uniq! { <span class="o">|</span>s<span class="o">|</span> s[<span class="n">0</span>] } <span class="c"># =&gt; [{&quot;student&quot;, &quot;sam&quot;}, {&quot;teacher&quot;, &quot;matz&quot;}]</span>
+<pre><code class="language-crystal">a <span class="o">=</span> [{<span class="s">&quot;student&quot;</span>, <span class="s">&quot;sam&quot;</span>, {<span class="s">&quot;student&quot;</span>, <span class="s">&quot;george&quot;</span>, {<span class="s">&quot;teacher&quot;</span>, <span class="s">&quot;matz&quot;</span>]
+a.uniq! { <span class="o">|</span>s<span class="o">|</span> s[<span class="n">0</span>]  <span class="c"># =&gt; [{&quot;student&quot;, &quot;sam&quot;}, {&quot;teacher&quot;, &quot;matz&quot;}]</span>
@@ -8489 +8489 @@
-b <span class="o">=</span> a.unstable_sort { <span class="o">|</span>a, b<span class="o">|</span> b <span class="o">&lt;=&gt;</span> a }
+b <span class="o">=</span> a.unstable_sort { <span class="o">|</span>a, b<span class="o">|</span> b <span class="o">&lt;=&gt;</span> a 
@@ -8551 +8551 @@
-a.unstable_sort! { <span class="o">|</span>a, b<span class="o">|</span> b <span class="o">&lt;=&gt;</span> a }
+a.unstable_sort! { <span class="o">|</span>a, b<span class="o">|</span> b <span class="o">&lt;=&gt;</span> a 
@@ -8581 +8581 @@
-b <span class="o">=</span> a.unstable_sort_by { <span class="o">|</span>word<span class="o">|</span> word.size }
+b <span class="o">=</span> a.unstable_sort_by { <span class="o">|</span>word<span class="o">|</span> word.size 
@@ -8611 +8611 @@
-a.usntable_sort_by! { <span class="o">|</span>word<span class="o">|</span> word.size }
+a.usntable_sort_by! { <span class="o">|</span>word<span class="o">|</span> word.size 
```

</details>